Improvements combobox component

chrisalat opened the pull request.

chrisalat committed:
(#6608) Resolved multiple bugfixes in combobox component

sdvg commented:
@chrisalat Beim Fokussieren der Vorschläge mit der Maus gibt es noch ein Problem:
![CleanShot 2024-07-15 at 12 52 40](https://github.com/user-attachments/assets/6645a493-21c1-4120-b6ea-0efc3a825b92)

In dem GIF mache ich folgendes:
- Eingabefeld fokussieren
- `a` eintippen
- Die Maus über dem ersten Vorschlag platzieren (nicht klicken)
- `Pfeiltaste nach oben` drücken

Ist: Der letzte Vorschlag ("Zentralafrikanische Republik") ist ausgewählt.
Soll: "Venezuela" ist ausgewählt.

sdvg commented on packages/components/src/components/combobox/shadow.tsx:
```diff
@@ -192,12 +195,17 @@ export class KolCombobox implements ComboboxAPI {
-								<span class={{ combobox__icon: true }}>
-									<KolIconTag _icons="codicon codicon-triangle-down" _label={translate('kol-dropdown')} onClick={this.toggleListbox.bind(this)} />
+								<span
+									role="button"
```
Ist die Änderung seitens Barrierefreiheit besprochen bzw. getestet? Ich dachte, es sollte explizit ein span ohne Rolle sein, sonst könnten wir ja auch eine `<button>` nehmen?

chrisalat replied:
Kann ich gerne Rückgängig machen. Wollte in diesem Fall nur die lint-Fehler lösen, da ich den onClick-Handler auf das Parent-Element verschoben habe. Das heißt die Info war nur einen span ohne Attribute zu verwenden?

sdvg replied:
> Das heißt die Info war nur einen span ohne Attribute zu verwenden?

Ich habe da keine weiteren Infos zu, glaube aber dass sich Cyrine bei der Initialentwicklung an einer funktionierenden Komponente orientiert hat und das von unserem Barrierefreiheitsteam mittlerweile so abgenommen wurde. Würde ohne Not also lieber keine Änderungen an am A11y-Tree machen.

sdvg commented on packages/components/src/components/combobox/shadow.tsx:
```diff
@@ -39,7 +39,7 @@ import { translate } from '../../i18n';
 export class KolCombobox implements ComboboxAPI {
 	@Element() private readonly host?: HTMLKolComboboxElement;
-	private ref?: HTMLSelectElement;
+	private ref?: HTMLInputElement;
```
```suggestion
	private refInput?: HTMLInputElement;
```

(Ich habe in einem anderen Branch schon angefangen, das zu vereinheitlichen, ist nur noch nicht gemerged. Dann weiß man auch, welcher `ref` für welches Element steht.)

sdvg commented on packages/components/src/components/combobox/shadow.tsx:
```diff
@@ -266,9 +274,16 @@ export class KolCombobox implements ComboboxAPI {
 				handleEvent(true, () => this.moveFocus(-1));
 				break;
 			}
+			case 'Tab':
```
Das funktioniert nur, wenn explizit Tab gedrückt wurde. Nicht für alle anderen Wege, wie der Fokus verloren gehen kann. Schau dir am besten mal an, wie das bei SingleSelect umgesetzt ist: https://github.com/public-ui/kolibri/pull/6546/files#diff-82183195a6f6176697e5362801d4f628dd0599c6f60ce3ee2ffce09a3ace4d71R636

isOpen würde ich immer auf `false` setzen, wenn der Fokus in irgendeiner Form verloren geht (einschließlich window focus). Warum wir hier `refInput.focus()` ausführen verstehe ich noch nicht.

chrisalat replied:
Ich verstehe den Ansatz einen Event-Listener zu verwenden und habe es auch versucht umzusetzen, aber das einzige Resultat welches ich damit hinbekomme, ist das, dass der Focus zu erst auf das Input zurückwandert und erst beim zweiten Tab das Dropdown schließt. `this.refInput?.focus();` sorgt derzeitig dafür das die Logik mit dem selben Tab das Input wieder fokussiert und gleichzeitig in das nächste Feld im DOM springt. Ohne diese Methode wandert der Fokus zu dem ersten Element im DOM.

chrisalat replied:
Wir hebeln ja die native Tab-Methode nicht aus, sondern erweitern diese nur

sdvg commented on packages/components/src/components/combobox/style.scss:
```diff
@@ -26,6 +26,13 @@
 
+		&__icon {
+			display: flex;
+			width: 44px;
```
```suggestion
			width: var(--a11y-min-size);
```

sdvg commented on packages/themes/default/src/components/combobox.scss:
```diff
@@ -1,5 +1,8 @@
-@import '../mixins/alert-wc.scss';
+@import '../mixins/alert-wc';
+
+$option-height: 3rem;
```
Bitte überall außer BMF immer rem() verwenden.

chrisalat replied:
Wollte ich eigentlich machen, aber die Konsole wirft mir einen Fehler, wenn ich rem in einer Variable verwende

sdvg replied:
Müssten wir trotzdem irgendwie lösen, sonst skaliert das im Zweifelsfall falsch. Sag Bescheid, wenn du hier Hilfe brauchst.

chrisalat committed:
(#6608) Review changes

chrisalat commented:
> @chrisalat Beim Fokussieren der Vorschläge mit der Maus gibt es noch ein Problem: ![CleanShot 2024-07-15 at 12 52 40](https://private-user-images.githubusercontent.com/971072/348679940-6645a493-21c1-4120-b6ea-0efc3a825b92.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEyMzE5MDcsIm5iZiI6MTcyMTIzMTYwNywicGF0aCI6Ii85NzEwNzIvMzQ4Njc5OTQwLTY2NDVhNDkzLTIxYzEtNDEyMC1iNmVhLTBlZmMzYTgyNWI5Mi5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxN1QxNTUzMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMTgwMTI5ZTU4OWQ3MDlkZWNlZTYyM2JlNGQ2OThmNGFmYjg3ZjUzYmJkYTViZTdkZDg4ODdiODI3MTQ1Y2JiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.QVr6NrYztqgn1_KhbsenmFT7hGRcU_I_dHg0misaai8)
> 
> In dem GIF mache ich folgendes:
> 
> * Eingabefeld fokussieren
> * `a` eintippen
> * Die Maus über dem ersten Vorschlag platzieren (nicht klicken)
> * `Pfeiltaste nach oben` drücken
> 
> Ist: Der letzte Vorschlag ("Zentralafrikanische Republik") ist ausgewählt. Soll: "Venezuela" ist ausgewählt.

Ich bin mir unsicher, ob das ein Fehler ist.
Der Browser selektiert das erste Element, da der Cursor auf diesem platziert ist. Wenn du den Cursor rechts neben das Dropdown verschiebst und die Pfeiltaste nach oben drückst, selektiert er die Zentralafrikanische Republik oder sollen wir den Cursor extra aushebeln, wenn die Tastatur verwendet wird?

sdvg commented:
> oder sollen wir den Cursor extra aushebeln, wenn die Tastatur verwendet wird?

Sollten wir, haben wir damals im Refinement so besprochen. (Ich weiß nicht, ob du da dabei warst.)

chrisalat committed:
(#6608) Review changes

chrisalat committed:
(#6608) Review changes

chrisalat committed:
(6608) Resolved merge conflicts

Refs: #6608

sdvg committed:
Combobox: Add method kolFocus

Refs: #6608

sdvg committed:
Combobox: Block mouse events on all handled keyboard shortcuts

Refs: #6608

sdvg committed:
Update Readme files

Refs: #6608

sdvg committed:
Refactor to use builtin event listeners

Refs: #6608

sdvg committed:
Refactor CSS: Move cursor to base style, fix BEM, remove unused styles

Refs: #6608

sdvg committed:
Combobox: handle blur

Refs: #6608

sdvg committed:
Remove unused code

Refs: #6608

sdvg committed:
Add DRAFT hint to component

Refs: #6608

sdvg committed:
Fix value for change events and associated form values

Refs: #6608

sdvg committed:
Merge branch 'develop' of github.com:public-ui/kolibri into 6608-combobox-optimization

* 'develop' of github.com:public-ui/kolibri:
  Remove KolSpan from component-list.ts
  fix: correct handling of invalid date inputs
  Remove KolSpan implementation
  (#6568) Review changes
  (#6568) Review changes
  cleanup
  Add more e2e test sample
  Move e2e tests to components folder
  Revert test change
  Test failing test run on CI
  Fix  playwright install
  Migrate Playwright tests to CI workflow
  Move workflow file
  Fix Safari tests
  Configure Playwright E2E tests with setContent pattern
  (#6568) Added label translations to table selection
  (#6568) Added selection control to table component and refactored some code

sdvg committed:
Fix combobox associated values

Refs: #6608

sdvg committed:
Fix broken forwardRef usage

Refs: #6608

sdvg committed:
Remove noise from console (use inputs-get-value) instead

Refs: #6608

sdvg committed:
Ignore ESLint error

Refs: #6608

sdvg committed:
Change arrow button to button element with tabindex -1

This should be a good compromise between an element that can become `document.activeElement` and an element that's not focusable (by keyboard).

Refs: #6608

sdvg commented on packages/components/src/components/combobox/shadow.tsx:
```diff
@@ -199,16 +205,19 @@ export class KolCombobox implements ComboboxAPI {
-								<span class={{ combobox__icon: true }}>
-									<KolIconTag _icons="codicon codicon-triangle-down" _label={translate('kol-dropdown')} onClick={this.toggleListbox.bind(this)} />
-								</span>
+								<button tabindex="-1" class="combobox__icon" onClick={this.toggleListbox.bind(this)}>
```
Button mit `tabindex=-1` scheint mir hier die ideale Lösung zu sein. Wir haben die Semantik eines Buttons und der Button kann das `activeElement` werden (Wichtig für den Check im `focusout`-Eventlistener). Gleichzeitig nimmt er aber keine "Tab-Position" ein, kann also nicht mit der Tastatur fokussiert werden.

sdvg commented on packages/samples/react/src/components/combobox/partials/cases.tsx:
```diff
@@ -1,28 +1,15 @@
 import { COUNTRY_SUGGESTIONS } from '../../../shares/country';
 import type { Components } from '@public-ui/components';
 
-export const ComboboxCases = forwardRef<HTMLKolComboboxElement, Components.KolCombobox>(function InputComboboxCases(props) {
```
Hier gab es einen Fehler in der Konsole, weil forwardRef nicht korrekt verwendet wurde. `forwardRef` schien mir überhaupt nicht notwendig, daher habe ich es ganz herausgenommen.

sdvg commented on packages/samples/react/src/components/combobox/partials/cases.tsx:
```diff
@@ -1,28 +1,15 @@
-				_label="Label"
-				_suggestions={COUNTRY_SUGGESTIONS}
-				_value={'Deutschland'}
-				_on={{
```
Die Logs kann ich wieder einfügen, wenn jemand darauf Wert legt. Ich fand sie beim Debuggen eher störend und die Events können wir bereits im inputs-get-value-Scenario testen.

sdvg committed:
Add missing <p> tag to description

Refs: #6608

sdvg committed:
Fix arrow button disabled state

Refs: #6608

sdvg committed:
Disable hover styles on disabled elements

Refs: #6608

sdvg committed:
Replace hardcoded rem values

Refs: #6608

sdvg committed:
Remove margin-bottom from combobox

Refs: #6608

sdvg committed:
Add missing sample routes

Refs: #6608

sdvg committed:
Update all snapshots$
$
Refs: #6608

sdvg committed:
Fix type import

Refs: #6608

sdvg committed:
Merge remote-tracking branch 'origin/6608-combobox-optimization' into 6608-combobox-optimization

* origin/6608-combobox-optimization:
  Update all snapshots$ $ Refs: #6608
